packages/lyra/tests/lyra.test.ts

```diff
@@ -364,9 +364,91 @@ describe("lyra", () => {
       tolerance: 3,
     });
 
+    // this now finds 3, which I think is correct?
     expect(moreTolerantSearch.count).toBe(2);
   });
 
+  it("Should tolerate typos better", async () => {
+    const db = new Lyra({
+      schema: {
+        txt: "string",
+      },
+      stemming: false // Disabling stemming to avoid confounding the results
+    })
+    
+    // Insert "stelle", and other words that are all within edit distance 2
+    await db.insert({ txt: 'stelle' })
+    await db.insert({ txt: 'stele' })
+    await db.insert({ txt: 'snelle' })
+    await db.insert({ txt: 'stellle' })
+    await db.insert({ txt: 'scelte' })
+    
+    let result = await db.search({ term: 'stelle', tolerance: 2 })
+    // this finds everything except 'scelte' but i believe this is due to the fact its comparing agianst the tokenized value, 'stell' and not 'stelle'
```

Comment:
The algorithm is incorrect: `findAllWords` checks the difference in length between the matched subtree term and the query, and bails out if it is larger than tolerance.

That's wrong, here's what happens with `scelte`: after the first `s` is matched, the recursive call with calculate the difference in length between `s` (the common prefix) and `scelte`, which is 5. Since it is above the tolerance, it bails out.

The fact that the difference in length is higher than the tolerance should only be leveraged on terminal nodes, not on prefixes: two long words can differ only by the first character.

That said, this specific algorithm is inefficient, as making it correct involves calculating the Levenshtein distance agains the query for the whole trie. But if you seek to get correctness first, then optimize, removing the condition on the difference in length is one step.

Reply:
@lucaong you're totally right, and we've already discussed this issue: https://github.com/nearform/lyra/pull/11#issuecomment-1184064203


> In a future release, as an improvement, we should aim at running the Levenshtein algorithm against the tree instead of words.
> Let me give you an example:
> If you're searching for "hello", but you mistype and write "hllo" instead, the current algorithm will work just fine.
> If you're searching for "hello", but you mistype and write "yellow" (with a tolerance of 2), this won't work, as we discard words with y as a prefix.
> This is something to consider for future releases, as for now, I am happy with the current PR

Eventually, we will need to calculate the edit distance on the tree rather than individual words

Reply:
I see, I missed the discussion on that PR 👍